packages/flutter/lib/src/widgets/heroes.dart

```diff
@@ -849,63 +882,89 @@ class HeroController extends NavigatorObserver {
-
-        final _HeroFlightManifest manifest = _HeroFlightManifest(
-          type: flightType,
-          overlay: navigator!.overlay,
-          navigatorRect: navigatorRect,
-          fromRoute: from,
-          toRoute: to,
-          fromHero: fromHeroes[tag]!,
-          toHero: toHeroes[tag]!,
-          createRectTween: createRectTween,
-          shuttleBuilder:
-              toShuttleBuilder ?? fromShuttleBuilder ?? _defaultHeroFlightShuttleBuilder,
-          isUserGestureTransition: isUserGestureTransition,
-          isDiverted: isDiverted,
-        );
+    final NavigatorState? navigator = this.navigator;
+    final OverlayState? overlay = navigator?.overlay;
+    // If the navigator or the overlay was removed before this end-of-frame
+    // callback was called, then don't actually start a transition, and we don'
+    // t have to worry about any Hero widget we might have hidden in a previous
+    // flight, or onging flights.
+    if (navigator == null || overlay == null)
+      return;
 
-        if (isDiverted)
-          _flights[tag]!.divert(manifest);
-        else
+    final RenderObject? navigatorRenderObject = navigator.context.findRenderObject();
+
+    if (navigatorRenderObject is! RenderBox) {
+      assert(false, 'Navigator $navigator has an invalid RenderObject type ${navigatorRenderObject.runtimeType}.');
+      return;
+    }
+    assert(navigatorRenderObject.hasSize);
+
+    // At this point, the toHeroes may have been built and laid out for the first time.
+    //
+    // If `fromSubtreeContext` is null, call endFlight on all toHeroes, for good measure.
+    // If `toSubtreeContext` is null abort existingFlights.
+    final BuildContext? fromSubtreeContext = from.subtreeContext;
+    final Map<Object, _HeroState> fromHeroes = fromSubtreeContext != null
+      ? Hero._allHeroesFor(fromSubtreeContext, isUserGestureTransition, navigator)
+      : const <Object, _HeroState>{};
+    final BuildContext? toSubtreeContext = to.subtreeContext;
+    final Map<Object, _HeroState> toHeroes = toSubtreeContext != null
+      ? Hero._allHeroesFor(toSubtreeContext, isUserGestureTransition, navigator)
+      : const <Object, _HeroState>{};
+
+    for (final MapEntry<Object, _HeroState> fromHeroEntry in fromHeroes.entries) {
+      final Object tag = fromHeroEntry.key;
+      final _HeroState fromHero = fromHeroEntry.value;
+      final _HeroState? toHero = toHeroes[tag];
+      final _HeroFlight? existingFlight = _flights[tag];
+      final _HeroFlightManifest? manifest = toHero == null
+        ? null
+        : _HeroFlightManifest(
+            type: flightType,
+            overlay: overlay,
+            navigatorSize: navigatorRenderObject.size,
+            fromRoute: from,
+            toRoute: to,
+            fromHero: fromHero,
+            toHero: toHero,
+            createRectTween: createRectTween,
+            shuttleBuilder: fromHero.widget.flightShuttleBuilder
```

Comment:
It looks like this should have been 
```
shuttleBuilder: toHero.widget.flightShuttleBuilder 
  ?? fromHero.widget.flightShuttleBuilder
  ?? _defaultHeroFlightShuttleBuilder,
```
That's the way it was, and the way it was documented. We must be missing a regression test for this case as well.